pkg/reconciler/revision/background.go

```diff
@@ -0,0 +1,260 @@
+		enqueue:  enqueue,
+
+		results: make(map[types.NamespacedName]*resolveResult),
+		queue:   workqueue.NewNamed("digests"),
+	}
+
+	return r
+}
+
+// Start starts the worker threads and runs maxInFlight workers until the stop
+// channel is closed. It returns a done channel which will be closed when all
+// workers have exited.
+func (r *backgroundResolver) Start(stop <-chan struct{}, maxInFlight int) (done chan struct{}) {
+	var wg sync.WaitGroup
+
+	// Run the worker threads.
+	wg.Add(maxInFlight)
+	for i := 0; i < maxInFlight; i++ {
+		go func() {
+			defer wg.Done()
+			for {
+				item, shutdown := r.queue.Get()
+				if shutdown {
+					return
+				}
+
+				rrItem, ok := item.(*workItem)
+				if !ok {
+					r.logger.Fatalf("Unexpected work item type: want: %s, got: %s\n",
+						reflect.TypeOf(&workItem{}).Name(), reflect.TypeOf(item).Name())
+				}
+
+				r.processWorkItem(rrItem)
+			}
+		}()
+	}
+
+	// Shut down the queue once the stop channel is closed, this will cause all
+	// the workers to exit.
+	go func() {
+		<-stop
+		r.queue.ShutDown()
+	}()
+
+	// Return a done channel which is closed once all workers exit.
+	done = make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+
+	return done
+}
+
+// Resolve is intended to be called from a reconciler to resolve a revision. If
+// the resolver already has the digest in cache it is returned immediately, if
+// it does not and no resolution is already in flight a resolution is triggered
+// in the background.
+// If this method returns nil, nil this implies a resolve was triggered or is
+// already in progress, so the reconciler should exit and wait for the revision
+// to be re-enqueued when the result is ready.
+func (r *backgroundResolver) Resolve(rev *v1.Revision, opt k8schain.Options, registriesToSkip sets.String, timeout time.Duration) ([]v1.ContainerStatus, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	name := types.NamespacedName{
+		Name:      rev.Name,
+		Namespace: rev.Namespace,
+	}
+
+	result, inFlight := r.results[name]
+	if !inFlight {
+		r.addWorkItems(rev, opt, registriesToSkip, timeout)
+		return nil, nil
+	}
+
+	if !result.ready() {
+		return nil, nil
+	}
+
+	return r.results[name].statuses, r.results[name].err
+}
+
+// addWorkItems adds a digest resolve item to the queue for each container in the revision.
+// This is expected to be called with the mutex locked.
+func (r *backgroundResolver) addWorkItems(rev *v1.Revision, opt k8schain.Options, registriesToSkip sets.String, timeout time.Duration) {
+	name := types.NamespacedName{
+		Name:      rev.Name,
+		Namespace: rev.Namespace,
+	}
+
+	r.results[name] = &resolveResult{
+		statuses:  make([]v1.ContainerStatus, len(rev.Spec.Containers)),
+		remaining: len(rev.Spec.Containers),
+		completionCallback: func() {
+			r.enqueue(name)
+		},
+	}
+
+	for i := range rev.Spec.Containers {
+		image := rev.Spec.Containers[i].Image
+
+		r.queue.Add(&workItem{
+			result:  r.results[name],
+			timeout: timeout,
```

Comment:
Do we need to pass this per digest, i.e. make it mutable? Seems like immutable would work here just fine too.

Reply:
I _think_ we need it to be per-digest because we want the timeout to be from when we actually start the download, e.g. if the thing is queued for 10 seconds that shouldn't count in the timeout. Also I want it passed via the Resolve() so that we can have it configurable via config map later.